Documentation/proposals/202304-agent-deployment-pattern-sidecar.md

```diff
@@ -0,0 +1,151 @@
+	if err != nil {
+		// We don't block pod creation if not able to generate patches
+		// Otherwise we could break the whole cluster.
+		level.Error(a.logger).Log("msg", "cannot generate patch for sidecar injection", "err", err)
+		return reviewResponse
+	}
+
+	pt := v1.PatchTypeJSONPatch
+	reviewResponse.PatchType = &pt
+	reviewResponse.Patch = []byte(fmt.Sprintf("[%s]", strings.Join(patches, ",")))
+	return reviewResponse
+}
+
+// shoulPodBeInjected checks if a given pod is selected by  PrometheusAgents through
+// ServiceMonitor/PodMonitor/Probe selectors.
+// If it should be injected, it returns the agents responsible for the injection.
+func shoulPodBeInjected(pod corev1.Pod) ([]monitoringv1alpha1.PrometheusAgent, error) {}
+
+// generatePatchesForSidecarInjection generates the patches necessary to inject the agent sidecar into the container
+// The agent parameter is used to configure extra container flags, remote-write configs and alikes
+func generatePatchesForSidecarInjection(raw []bytes, agents []monitoringv1alpha1.PrometheusAgent)
+```
+
+### Caveats
+
+#### Config Hot Reload
+
+While there is no research regarding the config-reloader efficiency, this particular container might become problematic for huge-scale environments, since a single change to a `ServiceMonitor`, `PodMonitor` or `PrometheusAgent` CR can potentially trigger a hot reload of millions of sidecar containers
+
+#### WAL is not optimized for small environments
+
+Prometheus Write-Ahead-log(WAL) is stored as a sequence of numbered files with 128MiB each by default. This means that, by default, at least 128MiB is needed for running Prometheus Agent if we ignore every other part of Prometheus. Using a sidecar, we're optimizing for horizontal scale and 128MiB might be much more than necessary to store metrics from a single Pod.
+
+#### Lack of High-Availability setup
+
+With the problem that Prometheus is not optimized for very small environments, injecting 2 sidecars per Pod sounds like a big waste of resources. However, with only 1 sidecar HA Prometheus won't be an option.
+
+With that said, having an HA Prometheus in the traditional deployment pattern seems to be more critical than the sidecar approach. That's because with Prometheus fails in the first approach we lose the monitoring stack for the whole cluster, while with the latter we just lose metrics from a pod.
+
+## Alternatives
+
+A common pattern used with Prometheus's Kubernetes service discovery is the [usage of annotation to declaratively tell Prometheus which endpoints need to be scraped](https://www.acagroup.be/en/blog/auto-discovery-of-kubernetes-endpoint-services-prometheus/). From a [code search at Github](https://github.com/search?q=prometheus.io%2Fscrape%3A+%22true%22&type=code) for `prometheus.io/scrape: "true"`, we can tell that this approach has good adoption already. To not conflict with the already commonly used annotation, we can start with our own, but with a very similar approach.
```

Comment:
I'm not sure to understand how this is an alternative?

Reply:
Instead of using ServiceMonitor/PodMonitor to select pods that need to be injected, we could look for annotations within Pods.

I've added extra info to the proposal